File: card.go
package main

import (
	"math/rand"
	"time"
)

type CardSuit string
type CardType string
type CardSpecial string

const (
	SuitPine          CardSuit = "Pine"
	SuitPlumBlossom            = "Plum Blossom"
	SuitCherryBlossom          = "Cherry Blossom"
	SuitWisteria               = "Wisteria"
	SuitIris                   = "Iris"
	SuitPeony                  = "Peony"
	SuitClover                 = "Clover"
	SuitPampas                 = "Pampas"
	SuitChrysanthemum          = "Chrysanthemum"
	SuitMaple                  = "Maple"
	SuitWillow                 = "Willow"
	SuitPaulownia              = "Paulownia"

	TypePlain  CardType = "Plain"
	TypeRibbon          = "Ribbon"
	TypeAnimal          = "Animal"
	TypeBright          = "Bright"

	SpecialBoar         CardSpecial = "Boar"
	SpecialDeer                     = "Deer"
	SpecialButterfly                = "Butterfly"
	SpecialSakeCup                  = "Sake Cup"
	SpecialRainMan                  = "Rain Man"
	SpecialPoetryRibbon             = "Poetry Ribbon"
	SpecialBlueRibbon               = "Blue Ribbon"
)

// Card represents a card
type Card struct {
	ID      int         `json:"id"`
	Suit    CardSuit    `json:"suit"`
	Type    CardType    `json:"type"`
	Special CardSpecial `json:"special,omitempty"`
}

// AllCards contains the information of every card
var AllCards = []Card{
	Card{
		ID:   0,
		Suit: SuitPine,
		Type: TypePlain,
	},
	Card{
		ID:   1,
		Suit: SuitPine,
		Type: TypePlain,
	},
	Card{
		ID:      2,
		Suit:    SuitPine,
		Type:    TypeRibbon,
		Special: SpecialPoetryRibbon,
	},
	Card{
		ID:   3,
		Suit: SuitPine,
		Type: TypeBright,
	},
	Card{
		ID:   4,
		Suit: SuitPlumBlossom,
		Type: TypePlain,
	},
	Card{
		ID:   5,
		Suit: SuitPlumBlossom,
		Type: TypePlain,
	},
	Card{
		ID:      6,
		Suit:    SuitPlumBlossom,
		Type:    TypeRibbon,
		Special: SpecialPoetryRibbon,
	},
	Card{
		ID:   7,
		Suit: SuitPlumBlossom,
		Type: TypeAnimal,
	},
	Card{
		ID:   8,
		Suit: SuitCherryBlossom,
		Type: TypePlain,
	},
	Card{
		ID:   9,
		Suit: SuitCherryBlossom,
		Type: TypePlain,
	},
	Card{
		ID:      10,
		Suit:    SuitCherryBlossom,
		Type:    TypeRibbon,
		Special: SpecialPoetryRibbon,
	},
	Card{
		ID:   11,
		Suit: SuitCherryBlossom,
		Type: TypeBright,
	},
	Card{
		ID:   12,
		Suit: SuitWisteria,
		Type: TypePlain,
	},
	Card{
		ID:   13,
		Suit: SuitWisteria,
		Type: TypePlain,
	},
	Card{
		ID:   14,
		Suit: SuitWisteria,
		Type: TypeRibbon,
	},
	Card{
		ID:   15,
		Suit: SuitWisteria,
		Type: TypeAnimal,
	},
	Card{
		ID:   16,
		Suit: SuitIris,
		Type: TypePlain,
	},
	Card{
		ID:   17,
		Suit: SuitIris,
		Type: TypePlain,
	},
	Card{
		ID:   18,
		Suit: SuitIris,
		Type: TypeRibbon,
	},
	Card{
		ID:   19,
		Suit: SuitIris,
		Type: TypeAnimal,
	},
	Card{
		ID:   20,
		Suit: SuitPeony,
		Type: TypePlain,
	},
	Card{
		ID:   21,
		Suit: SuitPeony,
		Type: TypePlain,
	},
	Card{
		ID:      22,
		Suit:    SuitPeony,
		Type:    TypeRibbon,
		Special: SpecialBlueRibbon,
	},
	Card{
		ID:      23,
		Suit:    SuitPeony,
		Type:    TypeAnimal,
		Special: SpecialButterfly,
	},
	Card{
		ID:   24,
		Suit: SuitClover,
		Type: TypePlain,
	},
	Card{
		ID:   25,
		Suit: SuitClover,
		Type: TypePlain,
	},
	Card{
		ID:   26,
		Suit: SuitClover,
		Type: TypeRibbon,
	},
	Card{
		ID:      27,
		Suit:    SuitClover,
		Type:    TypeAnimal,
		Special: SpecialBoar,
	},
	Card{
		ID:   28,
		Suit: SuitPampas,
		Type: TypePlain,
	},
	Card{
		ID:   29,
		Suit: SuitPampas,
		Type: TypePlain,
	},
	Card{
		ID:   30,
		Suit: SuitPampas,
		Type: TypeAnimal,
	},
	Card{
		ID:   31,
		Suit: SuitPampas,
		Type: TypeBright,
	},
	Card{
		ID:   32,
		Suit: SuitChrysanthemum,
		Type: TypePlain,
	},
	Card{
		ID:   33,
		Suit: SuitChrysanthemum,
		Type: TypePlain,
	},
	Card{
		ID:      34,
		Suit:    SuitChrysanthemum,
		Type:    TypeRibbon,
		Special: SpecialBlueRibbon,
	},
	Card{
		ID:      35,
		Suit:    SuitChrysanthemum,
		Type:    TypeAnimal,
		Special: SpecialSakeCup,
	},
	Card{
		ID:   36,
		Suit: SuitMaple,
		Type: TypePlain,
	},
	Card{
		ID:   37,
		Suit: SuitMaple,
		Type: TypePlain,
	},
	Card{
		ID:      38,
		Suit:    SuitMaple,
		Type:    TypeRibbon,
		Special: SpecialBlueRibbon,
	},
	Card{
		ID:      39,
		Suit:    SuitMaple,
		Type:    TypeAnimal,
		Special: SpecialDeer,
	},
	Card{
		ID:   40,
		Suit: SuitWillow,
		Type: TypePlain,
	},
	Card{
		ID:   41,
		Suit: SuitWillow,
		Type: TypeRibbon,
	},
	Card{
		ID:   42,
		Suit: SuitWillow,
		Type: TypeAnimal,
	},
	Card{
		ID:      43,
		Suit:    SuitWillow,
		Type:    TypeBright,
		Special: SpecialRainMan,
	},
	Card{
		ID:   44,
		Suit: SuitPaulownia,
		Type: TypePlain,
	},
	Card{
		ID:   45,
		Suit: SuitPaulownia,
		Type: TypePlain,
	},
	Card{
		ID:   46,
		Suit: SuitPaulownia,
		Type: TypePlain,
	},
	Card{
		ID:   47,
		Suit: SuitPaulownia,
		Type: TypeBright,
	},
}

func NewDeck() []Card {
	r := rand.New(rand.NewSource(time.Now().Unix()))
	deck := make([]Card, len(AllCards))
	perm := r.Perm(len(deck))
	for i, j := range perm {
		deck[i] = AllCards[j]
	}
	return deck
}

File: card.py

import copy
from enum import Enum
import logging
import random

class CardSuit(str, Enum):
    PINE            = "Pine"
    PLUM_BLOSSOM    = "Plum Blossom"
    CHERRY_BLOSSOM  = "Cherry Blossom"
    WISTERIA        = "Wisteria"
    IRIS            = "Iris"
    PEONY           = "Peony"
    CLOVER          = "Clover"
    PAMPAS          = "Pampas"
    CHRYSANTHEMUM   = "Chrysanthemum"
    MAPLE           = "Maple"
    WILLOW          = "Willow"
    PAULOWNIA       = "Paulownia"

class CardCategory(str, Enum):
    PLAIN   = "Plain"
    RIBBON  = "Ribbon"
    ANIMAL  = "Animal"
    BRIGHT  = "Bright"

class CardSpecial(str, Enum):
    NONE            = ""
    BOAR            = "Boar"
    DEER            = "Deer"
    BUTTERFLY       = "Butterfly"
    SAKE_CUP        = "Sake Cup"
    RAIN_MAIN       = "Rain Man"
    POETRY_RIBBON   = "Poetry Ribbon"
    BLUE_RIBBON     = "Blue Ribbon"
    YOKAI           = "Yokai"

class Card:
    def __init__(self, id, suit, category, special=CardSpecial.NONE):
        self.id = id
        self.suit = suit
        self.category = category
        self.special = special

    def toDict(self):
        return {
            'id': self.id,
            'suit': self.suit,
            'category': self.category,
            'special': self.special,
        }

all_cards = [
    Card(0,  CardSuit.PINE, CardCategory.PLAIN),
    Card(1,  CardSuit.PINE, CardCategory.PLAIN),
    Card(2,  CardSuit.PINE, CardCategory.RIBBON, CardSpecial.POETRY_RIBBON),
    Card(3,  CardSuit.PINE, CardCategory.BRIGHT),

    Card(4,  CardSuit.PLUM_BLOSSOM, CardCategory.PLAIN),
    Card(5,  CardSuit.PLUM_BLOSSOM, CardCategory.PLAIN),
    Card(6,  CardSuit.PLUM_BLOSSOM, CardCategory.RIBBON, CardSpecial.POETRY_RIBBON),
    Card(7,  CardSuit.PLUM_BLOSSOM, CardCategory.ANIMAL),

    Card(8,  CardSuit.CHERRY_BLOSSOM, CardCategory.PLAIN),
    Card(9,  CardSuit.CHERRY_BLOSSOM, CardCategory.PLAIN),
    Card(10, CardSuit.CHERRY_BLOSSOM, CardCategory.RIBBON, CardSpecial.POETRY_RIBBON),
    Card(11, CardSuit.CHERRY_BLOSSOM, CardCategory.BRIGHT),

    Card(12, CardSuit.WISTERIA, CardCategory.PLAIN),
    Card(13, CardSuit.WISTERIA, CardCategory.PLAIN),
    Card(14, CardSuit.WISTERIA, CardCategory.RIBBON),
    Card(15, CardSuit.WISTERIA, CardCategory.ANIMAL),
    
    Card(16, CardSuit.IRIS, CardCategory.PLAIN),
    Card(17, CardSuit.IRIS, CardCategory.PLAIN),
    Card(18, CardSuit.IRIS, CardCategory.RIBBON),
    Card(19, CardSuit.IRIS, CardCategory.ANIMAL, CardSpecial.YOKAI),
    
    Card(20, CardSuit.PEONY, CardCategory.PLAIN),
    Card(21, CardSuit.PEONY, CardCategory.PLAIN),
    Card(22, CardSuit.PEONY, CardCategory.RIBBON, CardSpecial.BLUE_RIBBON),
    Card(23, CardSuit.PEONY, CardCategory.ANIMAL, CardSpecial.BUTTERFLY),
    
    Card(24, CardSuit.CLOVER, CardCategory.PLAIN),
    Card(25, CardSuit.CLOVER, CardCategory.PLAIN),
    Card(26, CardSuit.CLOVER, CardCategory.RIBBON),
    Card(27, CardSuit.CLOVER, CardCategory.ANIMAL, CardSpecial.BOAR),
    
    Card(28, CardSuit.PAMPAS, CardCategory.PLAIN),
    Card(29, CardSuit.PAMPAS, CardCategory.PLAIN),
    Card(30, CardSuit.PAMPAS, CardCategory.ANIMAL),
    Card(31, CardSuit.PAMPAS, CardCategory.BRIGHT),
    
    Card(32, CardSuit.CHRYSANTHEMUM, CardCategory.PLAIN),
    Card(33, CardSuit.CHRYSANTHEMUM, CardCategory.PLAIN),
    Card(34, CardSuit.CHRYSANTHEMUM, CardCategory.RIBBON, CardSpecial.BLUE_RIBBON),
    Card(35, CardSuit.CHRYSANTHEMUM, CardCategory.ANIMAL, CardSpecial.SAKE_CUP),
    
    Card(36, CardSuit.MAPLE, CardCategory.PLAIN),
    Card(37, CardSuit.MAPLE, CardCategory.PLAIN),
    Card(38, CardSuit.MAPLE, CardCategory.RIBBON, CardSpecial.BLUE_RIBBON),
    Card(39, CardSuit.MAPLE, CardCategory.ANIMAL, CardSpecial.DEER),
    
    Card(40, CardSuit.WILLOW, CardCategory.PLAIN, CardSpecial.YOKAI),
    Card(41, CardSuit.WILLOW, CardCategory.RIBBON),
    Card(42, CardSuit.WILLOW, CardCategory.ANIMAL),
    Card(43, CardSuit.WILLOW, CardCategory.BRIGHT, CardSpecial.RAIN_MAIN),
    
    Card(44, CardSuit.PAULOWNIA, CardCategory.PLAIN),
    Card(45, CardSuit.PAULOWNIA, CardCategory.PLAIN),
    Card(46, CardSuit.PAULOWNIA, CardCategory.PLAIN, CardSpecial.YOKAI),
    Card(47, CardSuit.PAULOWNIA, CardCategory.BRIGHT),
]

def new_deck():
    deck = copy.deepcopy(all_cards)
    random.shuffle(deck)
    return deck